File: dr_py/ffys.py
# -*- coding: utf-8 -*-
"""
TP‑3 Spider脚本，type=3，TP3引擎专用
站点：ffys.fun
现象：分类正常展示，列表无数据
TP3禁止使用urllib，全部使用引擎内置http()
分类ID：20电影,21剧集,22动漫,23综艺
"""
class Spider:
    HOST = "https://www.ffys.fun"
    UA = "Mozilla/5.0 (Android; Mobile) Chrome/130.0.0.0 Safari/537.36"

    def getName(self):
        return "飞飞影视"

    def init(self, extend=""):
        self.header = {
            "User‑Agent": self.UA,
            "Referer": self.HOST + "/",
            "Accept": "application/json;charset=utf‑8"
        }

    def homeContent(self, filter):
        result = {
            "class": [
                {"type_name": "电影", "type_id": "20"},
                {"type_name": "剧集", "type_id": "21"},
                {"type_name": "动漫", "type_id": "22"},
                {"type_name": "综艺", "type_id": "23"},
            ],
            "filters": {},
            "list": []
        }
        # TP3内置http请求函数
        resp = self.http(f"{self.HOST}/api/home", self.header)
        if resp["code"] != 200:
            return result
        try:
            js = self.jsonLoads(resp["content"])
            data_list = js.get("data", [])
            for item in data_list:
                result["list"].append({
                    "vod_id": str(item.get("id")),
                    "vod_name": item.get("title", ""),
                    "vod_pic": self.fixImg(item.get("pic", "")),
                    "vod_remarks": item.get("update", "")
                })
        except Exception:
            pass
        return result

    def homeVideoContent(self):
        result = {"list": []}
        resp = self.http(f"{self.HOST}/api/home", self.header)
        if resp["code"] != 200:
            return result
        try:
            js = self.jsonLoads(resp["content"])
            arr = js.get("data", [])
            for item in arr[:24]:
                result["list"].append({
                    "vod_id": str(item.get("id")),
                    "vod_name": item.get("title", ""),
                    "vod_pic": self.fixImg(item.get("pic", "")),
                    "vod_remarks": item.get("update", "")
                })
        except Exception:
            pass
        return result

    def categoryContent(self, tid, pg, filter, extend):
        result = {
            "list": [],
            "page": int(pg),
            "pagecount": 9999,
            "limit": 30,
            "total": 9999
        }
        api_url = f"{self.HOST}/api/cate?type={tid}&page={pg}"
        resp = self.http(api_url, self.header)
        if resp["code"] != 200:
            return result
        try:
            js = self.jsonLoads(resp["content"])
            data_list = js.get("data", [])
            for item in data_list:
                result["list"].append({
                    "vod_id": str(item.get("id")),
                    "vod_name": item.get("title", ""),
                    "vod_pic": self.fixImg(item.get("pic", "")),
                    "vod_remarks": item.get("update", "")
                })
        except Exception:
            pass
        return result

    def detailContent(self, ids):
        result = {"list": []}
        vid = ids[0]
        resp = self.http(f"{self.HOST}/api/detail?id={vid}", self.header)
        if resp["code"] != 200:
            return result
        try:
            js = self.jsonLoads(resp["content"])
            d = js["data"]
            vod = {
                "vod_id": str(d["id"]),
                "vod_name": d.get("title", ""),
                "vod_pic": self.fixImg(d.get("pic", "")),
                "vod_year": d.get("year", ""),
                "vod_area": d.get("area", ""),
                "vod_actor": d.get("actor", ""),
                "vod_director": d.get("director", ""),
                "vod_content": d.get("desc", ""),
                "vod_remarks": d.get("update", ""),
                "vod_play_from": "",
                "vod_play_url": ""
            }
            froms = []
            urls = []
            playList = d.get("playList", [])
            for idx, pl in enumerate(playList):
                ep_array = []
                for ep in pl.get("list", []):
                    ep_array.append(f"{ep['name']}${ep['url']}")
                froms.append(pl.get("name", f"线路{idx+1}"))
                urls.append("#".join(ep_array))
            vod["vod_play_from"] = "$$$".join(froms)
            vod["vod_play_url"] = "$$$".join(urls)
            result["list"].append(vod)
        except Exception:
            pass
        return result

    def searchContent(self, key, quick):
        result = {"list": []}
        from urllib.parse import quote
        url = f'{self.HOST}/api/search?wd={quote(key)}&page=1'
        resp = self.http(url, self.header)
        if resp["code"] != 200:
            return result
        try:
            js = self.jsonLoads(resp["content"])
            arr = js.get("data", [])
            for item in arr:
                result["list"].append({
                    "vod_id": str(item.get("id")),
                    "vod_name": item.get("title", ""),
                    "vod_pic": self.fixImg(item.get("pic", "")),
                    "vod_remarks": item.get("update", "")
                })
        except Exception:
            pass
        return result

    def playerContent(self, flag, id, vipFlags):
        return {
            "parse": 0,
            "jx": 0,
            "url": id,
            "header": self.header
        }

    def fixImg(self, url):
        if not url:
            return ""
        url = url.strip()
        if url.startswith("http"):
            return url
        if url.startswith("//"):
            return "https:" + url
        return self.HOST + url

File: dr_py/ffys.js
/**
 * @title 飞飞影视
 * @type 0
 * @api https://www.ffys.fun
 */

const UA = "Mozilla/5.0 (Android TV) AppleWebKit/537.36 Chrome/126.0.0.0 Safari/537.36";
const siteUrl = "https://www.ffys.fun";

const cateMap = [
    { typeId: 1, typeName: "电影" },
    { typeId: 2, typeName: "剧集" },
    { typeId: 3, typeName: "动漫" },
    { typeId: 4, typeName: "综艺" },
    { typeId: 6, typeName: "爽剧" }
];

async function home() {
    let list = [];
    let res = await fetch(`${siteUrl}/api/home/recommend`, {
        headers: { "User-Agent": UA, "Referer": siteUrl }
    });
    let json = await res.json();
    if (json?.data) {
        list.push({
            title: "🔥首页推荐",
            data: json.data.map(item => formatVod(item))
        });
    }
    for (let c of cateMap) {
        let r = await fetch(`${siteUrl}/api/video/list?type=${c.typeId}&page=1&limit=12`, {
            headers: { "User-Agent": UA, "Referer": siteUrl }
        });
        let j = await r.json();
        if (j?.data?.list) {
            list.push({
                title: c.typeName,
                data: j.data.list.map(item => formatVod(item))
            });
        }
    }
    return setHome(list);
}

async function category(tid, pg) {
    let cid = Number(tid);
    let limit = 20;
    let res = await fetch(`${siteUrl}/api/video/list?type=${cid}&page=${pg}&limit=${limit}`, {
        headers: {
            "User-Agent": UA,
            "Referer": siteUrl
        }
    });
    let json = await res.json();
    let arr = [];
    if (json?.data?.list) {
        arr = json.data.list.map(item => formatVod(item));
    }
    return setVod(arr, pg < json?.data?.totalPage);
}

async function search(wd, pg) {
    let res = await fetch(`${siteUrl}/api/video/search?keyword=${encodeURIComponent(wd)}&page=${pg}&limit=20`, {
        headers: {
            "User-Agent": UA,
            "Referer": siteUrl
        }
    });
    let json = await res.json();
    let arr = [];
    if (json?.data?.list) {
        arr = json.data.list.map(item => formatVod(item));
    }
    return setVod(arr, pg < json?.data?.totalPage);
}

async function detail(id) {
    let res = await fetch(`${siteUrl}/api/video/detail?id=${id}`, {
        headers: {
            "User-Agent": UA,
            "Referer": siteUrl
        }
    });
    let json = await res.json();
    if (!json?.data) return setDetail({});
    let info = json.data;
    let vod = formatVod(info);

    let playList = [];
    if (info?.playList && Array.isArray(info.playList)) {
        for (let pl of info.playList) {
            let ep = [];
            for (let p of pl.playUrls) {
                ep.push(`${p.name}$${p.url}`);
            }
            playList.push(ep.join("#"));
        }
    }
    vod.vod_play_from = playList.map((_, i) => `线路${i + 1}`).join("$$$");
    vod.vod_play_url = playList.join("$$$");
    return setDetail(vod);
}

function formatVod(item) {
    return {
        vod_id: item.id,
        vod_name: item.title,
        vod_pic: item.image,
        vod_remarks: item.updateInfo || "",
        vod_year: item.year || "",
        vod_area: item.area || "",
        vod_class: item.typeName || "",
        vod_actor: item.actor || "",
        vod_director: item.director || "",
        vod_content: item.desc || ""
    };
}

async function getCategory() {
    return cateMap.map(x => {
        return {
            type_id: x.typeId,
            type_name: x.typeName
        };
    });
}
